### DISCUSSION TITLE
React router 7.0.0-pre.2 types for useFetcher and useActionData

### DISCUSSION THREAD

**Discussion by lasseklovstad** [remix-run/react-router#12192] [Q&A]
In the new prerelease I had some issue with the types return from useFetcher and useActionData when I return a redirect in the action function. I found it easier to throw the redirect because then the type returned from the redirect didn't get infered to the action data in useFetcher and useActionData.

```typescript
import type { ActionData, ActionArgs } from "./+types.Page";

export async function action({ request }: ActionArgs) {
  const isLoggedIn = getIsLoggedIn(request)
  if(loggedIn){
    // Do something
    return {ok: true, ...someMoreData }
  }
  return redirect("/login");
}

export default function Page(){
    // The type here includes headers and stuff from the redirect Response
    const actionData = useActionData<ActionData>();
}
```

Is it recommended to always throw the redirect to not have this issue with the types?

**Comment by melodyclue** [comment]
@lasseklovstad 
They say this is fixed in 7.1.0

https://github.com/remix-run/react-router/blob/main/CHANGELOG.md#v710
